goal_src/jak2/engine/physics/rigid-body.gc

```diff
@@ -5,5 +5,1010 @@
+;; og:ignore-form:collide-shape-moving-method-67
+
 ;; DECOMP BEGINS
 
+(deftype rigid-body-work (structure)
+  ((max-ang-momentum float  :offset-assert   0)
+   (max-ang-velocity float  :offset-assert   4)
+   )
+  :method-count-assert 9
+  :size-assert         #x8
+  :flag-assert         #x900000008
+  )
+
+
+(define *rigid-body-work* (new 'static 'rigid-body-work))
+
+(defmethod new rigid-body-control ((allocation symbol) (type-to-make type) (arg0 process))
+  (let ((v0-0 (object-new allocation type-to-make (the-as int (-> type-to-make size)))))
+    (set! (-> v0-0 process) arg0)
+    v0-0
+    )
+  )
+
+(defmethod relocate rigid-body-control ((obj rigid-body-control) (arg0 int))
+  (&+! (-> obj process) arg0)
+  obj
+  )
+
+(defmethod rigid-body-info-method-9 rigid-body-info ((obj rigid-body-info))
+  (let ((f24-0 (-> obj mass))
+        (f28-0 (-> obj inertial-tensor-box 0))
+        (f30-0 (-> obj inertial-tensor-box 1))
+        (f26-0 (-> obj inertial-tensor-box 2))
+        )
+    (let ((f0-0 f24-0))
+      (set! (-> obj inv-mass) (/ 1.0 f0-0))
+      )
+    (matrix-identity! (-> obj inertial-tensor))
+    (matrix-identity! (-> obj inv-inertial-tensor))
+    (let ((f0-4 (* 0.083333336 f24-0)))
+      (let* ((f1-1 f30-0)
+             (f1-3 (* f1-1 f1-1))
+             (f2-0 f26-0)
+             )
+        (set! (-> obj inertial-tensor vector 0 x) (* f0-4 (+ f1-3 (* f2-0 f2-0))))
+        )
+      (let ((f1-6 f28-0))
+        (set! (-> obj inertial-tensor vector 1 y) (* f0-4 (+ (* f1-6 f1-6) (* f26-0 f26-0))))
+        )
+      (set! (-> obj inertial-tensor vector 2 z) (* f0-4 (+ (* f28-0 f28-0) (* f30-0 f30-0))))
+      )
+    )
+  (let ((f0-6 (-> obj inertial-tensor vector 0 x)))
+    (set! (-> obj inv-inertial-tensor vector 0 x) (/ 1.0 f0-6))
+    )
+  (let ((f0-9 (-> obj inertial-tensor vector 1 y)))
+    (set! (-> obj inv-inertial-tensor vector 1 y) (/ 1.0 f0-9))
+    )
+  (let ((f0-12 (-> obj inertial-tensor vector 2 z)))
+    (set! (-> obj inv-inertial-tensor vector 2 z) (/ 1.0 f0-12))
+    )
+  0
+  (none)
+  )
+
+(defmethod clear-force-torque! rigid-body ((obj rigid-body))
+  (set! (-> obj force quad) (the-as uint128 0))
+  (set! (-> obj torque quad) (the-as uint128 0))
+  0
+  (none)
+  )
+
+(defmethod clear-momentum! rigid-body ((obj rigid-body))
+  (set! (-> obj lin-momentum quad) (the-as uint128 0))
+  (set! (-> obj ang-momentum quad) (the-as uint128 0))
+  0
+  (none)
+  )
+
+(defmethod rigid-body-method-24 rigid-body ((obj rigid-body))
+  (when #t
+    (quaternion->matrix (-> obj matrix) (-> obj rotation))
+    (let ((s5-0 (new 'stack-no-clear 'vector)))
+      (vector-rotate*! s5-0 (-> obj info cm-offset-joint) (-> obj matrix))
+      (vector-! (-> obj matrix trans) (-> obj position) s5-0)
+      )
+    )
+  0
+  (none)
+  )
+
+(defmethod rigid-body-method-26 rigid-body ((obj rigid-body) (arg0 vector) (arg1 quaternion))
+  (let ((s3-0 (new 'stack-no-clear 'rigid-body-impact)))
```

Comment:
I think this will cause problems: `(quaternion->matrix (the-as matrix (-> s3-0 normal)) arg1)` will write 16 bytes past the end of `s3-0`. If you put an inline array of 8 vectors, it will be ugly, but won't have stack corruption problems.

Reply:
Yeah this type was a total guess, I wonder if it was originally just a bunch of vectors or if there is a type im missing.